algebird-core/src/main/scala/com/twitter/algebird/CountMinSketch.scala

```diff
@@ -458,6 +458,59 @@ case class CMSItem[K](item: K, override val totalCount: Long, override val param
 }
 
 /**
+ * A sparse Count-Min sketch structure, used for situations where the key is highly skewed.
+ */
+case class SparseCMS[K](exactCountTable: Map[K, Long],
+  override val totalCount: Long,
+  override val params: CMSParams[K]) extends CMS[K](params) {
+
+  override def +(x: K, count: Long): CMS[K] = {
+    val currentCount = exactCountTable.getOrElse(x, 0L)
+    val newTable = exactCountTable.updated(x, currentCount + count)
+    if (newTable.size < width * depth) {
+      // still sparse
+      SparseCMS(newTable, totalCount + count, params)
+    } else {
+      // Create new CMSInstace
+      var cms = CMSInstance[K](params)
+      newTable.foreach {
+        case (x, count) =>
+          cms = cms + (x, count)
+      }
+      cms
+    }
+  }
+
+  override def ++(other: CMS[K]): CMS[K] = {
+    other match {
+      case other: CMSZero[_] => this
+      case other: CMSItem[K] => this + other.item
+      case other: SparseCMS[K] =>
+        val newTable = exactCountTable ++ other.exactCountTable
+        if (newTable.size < width * depth) {
+          // still sparse
+          SparseCMS(newTable, totalCount + other.totalCount, params)
+        } else {
+          // Create new CMSInstace
+          var cms = CMSInstance[K](params)
+          newTable.foreach {
+            case (x, count) =>
+              cms = cms + (x, count)
+          }
+          cms
+        }
+
+      case other: CMSInstance[K] => other ++ this
+    }
+  }
+
+  override def frequency(x: K): Approximate[Long] = Approximate.exact(exactCountTable.getOrElse(x, 0L))
+
+  override def innerProduct(other: CMS[K]): Approximate[Long] =
+    exactCountTable.map { case (x, count) => Approximate.exact(count) * other.frequency(x) }.reduce { _ + _ }
```

Comment:
let's do `exactCountTable.iterator.map ...` here so we are lazy until we hit the reduce. Also, what if exactCountTable is empty? I think reduce throws in that case. Might want to do .reduceOption and put a `getOrElse(Approximate.exact(0L))`